pysparkling/sql/expressions/mappers.py

```diff
@@ -75,6 +87,115 @@ def __str__(self):
+                pass
+        self.dtype = dtype
+        self.dtype_str = dtype
+        if self.dtype_str == 'string':
+            self.dtype = str
+        elif self.dtype_str == 'integer':
+            self.dtype = int
+        elif self.dtype_str == 'boolean':
+            self.dtype = bool
+        elif self.dtype_str in ['str', 'int', 'float', 'bool']:
+            self.dtype = eval(self.dtype_str)
+        else:
+            raise NotImplementedError("Unknown cast type: {}".format(dtype))
+
+    def eval(self, row, schema):
+        return self.dtype(self.e.eval(row, schema))
+
+    def __str__(self):
+        return "cast({0})".format(self.dtype_str)
+
+
+class CaseWhen(Expression):
+    def __init__(self, condition, function):
+        super().__init__(condition, function)
+
+        self.conditions = [condition]
+        self.functions = [function]
+        self.function_b = None
+
+    def otherwise(self, function):
+        self.function_b = function
+        return self
+
+    def when(self, condition, function):
+        self.conditions.append(condition)
+        self.functions.append(function)
+        return self
+
+    def eval(self, row, schema):
+        for condition, function in zip(self.conditions, self.functions):
+            trueth = condition.eval(row, schema)
+            if trueth:
+                return function
+                # self.function.eval(row, schema)
+        if self.function_b is not None:
+            return self.function_b
+            # self.function_b.eval(row, schema)
+
+    def __str__(self):
+        return "CASE "+" ".join(["WHEN {0} THEN {1}".format(condition, function)
+                                for condition, function in zip(self.conditions, self.functions)]) + \
+               (" ELSE {}".format(self.function_b) if self.function_b is not None else "") + " END"
+
+
+class RegExpExtract(Expression):
+    def __init__(self, e, exp, groupIdx):
+        super().__init__(e, exp, groupIdx)
+
+        import re
+        regexp = re.compile(exp)
+
+        def fn(x):
+            match = regexp.search(x)
+            ret = match.group(groupIdx)
+            return ret
+
+        self.fn = fn
+        self.exp = exp
+        self.groupIdx = groupIdx
+        self.e = e
+
+    def eval(self, row, schema):
+        return self.fn(self.e.eval(row, schema))
+
+    def __str__(self):
+        return "RegExpExtract({0}, {1})".format(self.exp, self.groupIdx)
+
+
+class RegExpReplace(Expression):
+    def __init__(self, e, exp, replacement):
+        super().__init__(e, exp, replacement)
+
+        import re
+        regexp = re.compile(exp)
+
+        def fn(x):
+            return regexp.sub(replacement, x)
+
+        self.fn = fn
+        self.exp = exp
+        self.replacement = replacement
+        self.e = e
+
+    def eval(self, row, schema):
+        return self.fn(self.e.eval(row, schema))
+
+    def __str__(self):
+        return "RegExpReplace({0}, {1})".format(self.exp, self.replacement)
```

Comment:
Same here, regexp_replace instead of RegExpReplace and self.e is missing

Reply:
yeap